src/wh_server.c

```diff
@@ -28,8 +30,613 @@
+    meta->len = WOLFHSM_KEYSIZE;
+    ret = hsmReadKey(ctx, meta, keyBuf);
+    /* decode the key */
+    if (ret == 0) {
+        keySz = meta->len / 3;
+        ret = wc_ecc_import_unsigned(key, keyBuf, keyBuf + keySz,
+            keyBuf + keySz * 2, curveId);
+    }
+    return ret;
+}
+#endif
+
+static int hsmGetUniqueId(whServerContext* server)
+{
+    int i;
+    int ret = 0;
+    uint16_t id;
+    /* try every index, unless both cache and nvm are full one will work */
+    for (id = 1; id < WOLFHSM_NUM_RAMKEYS + 1; id++) {
+        /* check against chache keys */
+        for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+            if (id == server->cache[i].meta->id)
+                break;
+        }
+        /* try again if match */
+        if (i < WOLFHSM_CACHE_COUNT)
+            continue;
+#if 0
+        /* check against nvm keys */
+        for (i = 0; i < WOLFHSM_KEYSLOT_COUNT && ret == 0; i++) {
+            if (ctx->nvmMetaCache[i].id == id) {
+                break;
+            }
+        }
+        /* try again if match */
+        if (i < WOLFHSM_KEYSLOT_COUNT)
+            continue;
+#endif
+        /* break if our id didn't match either cache or nvm */
+        break;
+    }
+    /* if we've run out of space, return MEMORY_E */
+    if (ret == 0 && id >= WOLFHSM_NUM_RAMKEYS + 1)
+        ret = MEMORY_E;
+    /* ultimately, return found id */
+    if (ret == 0)
+        ret = id;
+    return ret;
+}
+
+static int hsmCacheKey(whServerContext* server, NvmMetaData* meta, uint8_t* in)
+{
+    int i;
+    int foundIndex = -1;
+    /* make sure id is valid */
+    if (meta->id == 0)
+        return BAD_FUNC_ARG;
+    /* make sure the key fits in a slot */
+    if (meta->len > WOLFHSM_NVM_MAX_OBJECT_SIZE)
+        return BUFFER_E;
+    for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+        /* check for empty slot or rewrite slot */
+        if (foundIndex == -1 && (server->cache[i].meta->id == 0 ||
+            server->cache[i].meta->id == meta->id)) {
+            foundIndex = i;
+        }
+    }
+#if 0
+    /* if no empty slots, check for a commited key we can evict */
+    if (foundIndex == -1) {
+        for (i = 0; i < WOLFHSM_CACHE_COUNT; i++) {
+            if (ctx->cache[i].commited == 1) {
+                foundIndex = i;
+                break;
+            }
+        }
+    }
+    /* return error if we are out of cache slots */
+    if (foundIndex == -1)
+        return MEMORY_E;
+    ctx->cache[foundIndex].commited = 0;
+    /* check if the key is in nvm, mark as commited */
+    for (i = 0; i < WOLFHSM_KEYSLOT_COUNT; i++) {
+        /* check for duplicate id, shouldn't cache a key in nvm */
+        if (ctx->nvmMetaCache[i].id == meta->id)
+            ctx->cache[foundIndex].commited = 1;
+    }
+#endif
+    /* write key if slot found */
+    XMEMCPY((uint8_t*)server->cache[foundIndex].buffer, in, meta->len);
+    XMEMCPY((uint8_t*)server->cache[foundIndex].meta, (uint8_t*)meta,
+        sizeof(NvmMetaData));
```

Comment:
probably want `sizeof(server->cache[foundIndex].meta)`, especially since in my other comment I point out that we want to change the type to `whNvmMetaData`